Add utf8 support for string_split

yongtang opened the pull request:
This fix is an effort to try to address the request raised in #11399 where it was not possible to split utf8 strings with `tf.string_split`.

This fix adds an additional attr of `encoding` so that `utf8` could be specified for `string_split`.

This fix fixes #11399.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add utf8 support for string_split

This fix is an effort to try to address the request raised in
11399 where it was not possible to split utf8 strings with
`tf.string_split`.

This fix adds an additional attr of `encoding` so that
`utf8` could be specified for `string_split`.

This fix fixes 11399.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test cases for utf8 support for string_split.

This commit adds several test cases for utf8 support of
string_split.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

drpngx commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -28,19 +28,62 @@ namespace tensorflow {
 
-std::vector<string> Split(const string& str, const string& delimiter,
-                          const bool skipEmpty) {
+Status Split(std::vector<string>& char_vector, const string& str,
```
We don't do references. Instead, pass the output as the last parameter, as a pointer:

```Status Split(const string& str, const string& delimiter, ..., std::vector<string>* char_vector)
```

yongtang replied:
Thanks @drpngx. Done.

drpngx commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -76,6 +127,12 @@ class StringSplitOp : public OpKernel {
     // Empty delimiter means split the input character by character.
+    // Empty delimiter could be combined with a non-empty encoding.
+    if (encoding_ != "") {
+      OP_REQUIRES(ctx, delimiter == "",
```
Why not implement this case?

yongtang replied:
Done. Thanks.

drpngx commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -28,19 +28,62 @@ namespace tensorflow {
-  for (size_t i = 0; i < str.size(); ++i) {
-    char_vector[i] = str[i];
+  if (encoding == "utf8") {
+    // Bytes    Byte 1    Byte 2    Byte 3    Byte 4
```
Could we push this upstream to str_util and provide a variant there? It might be useful for the rest of the code. If moved there, it will also have to have a unit test on its own.

yongtang replied:
Done.

drpngx commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -28,19 +28,62 @@ namespace tensorflow {
+    for (size_t i = 0; i < str.size(); ++i) {
+      if (off == i) {
+        if ((str[i] & 0x80) == 0x00) {
+          char_vector.push_back(str.substr(off, 1));
```
emplace_back

yongtang replied:
Done.

drpngx commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -28,19 +28,62 @@ namespace tensorflow {
+          len = 4;
+        } else {
+          char_vector.clear();
+          return errors::Internal("invalid utf8 encoding");
```
maybe report at which character we failed

yongtang replied:
Done.

drpngx commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -28,19 +28,62 @@ namespace tensorflow {
+      }
+    }
+    if (off < str.size()) {
+      // Not enough chars
```
Remove comment and put that in the error message.

yongtang replied:
Done.

drpngx commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -28,19 +28,62 @@ namespace tensorflow {
+    }
+    if (off < str.size()) {
+      // Not enough chars
+      return errors::Internal("invalid utf8 encoding");
```
I guess these are not really internal errors, right? They could be due to invalid input?

yongtang replied:
Done.

drpngx reviewed (changes requested):
After we agree, this will require an API review.

yongtang committed:
Move SplitUTF8 to tensorflow/core/lib/strings/str_util.cc

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @drpngx for the review. The PR has been updated with comments addressed. Please take a look.

drpngx commented:
Jenkins, test this please.

On Sep 14, 2017 9:17 PM, "Yong Tang" <notifications@github.com> wrote:

> Thanks @drpngx <https://github.com/drpngx> for the review. The PR has
> been updated with comments addressed. Please take a look.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/12971#issuecomment-329676587>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AT_SbV-wjL_qPMRw1Jn7PG3S1fA7qRpGks5sifpUgaJpZM4PTdqy>
> .
>

yongtang commented:
One Jenkins failure is `//tensorflow/tools/api/tests:api_compatibility_test` which is expected (API Change).

Another Jenkins failure is `//tensorflow/python/kernel_tests:fft_ops_test`. I am wondering if it is untreated? (saw this failure in other PRs as well).

drpngx commented:
Oh you need to rebase, we've fixed that. The `fft_ops_test` is fixed internally, I will push that later.

drpngx commented on tensorflow/core/lib/strings/str_util.h:
```diff
@@ -131,6 +131,15 @@ std::vector<string> Split(StringPiece text, StringPiece delims);
+// one ASCII or one unicode char. If successful, the result
+// is passed to "*result" and returns true. Otherwise returns false
+// and error message is passed to "*error".
+bool SplitUTF8(StringPiece text, string delim, std::vector<string>* result,
```
Can we use the `Status` object to return errors?

Also, you should use `const string& delim`, or `StringPiece delim`. Isn't there a restriction that `delim` is only one character? If so, we should document that, and check for that in the implementation.

yongtang replied:
Thanks @drpngx. The PR has been updated with `Status` returned and `const string& delim`.

For `delim` length, in the kernel of StringSplitOp, there is already a check to guard the length of `delim`.

In the util library (`str_util.h`) the `SplitUTF8` actually allows a UTF8 encoded string (1-4 chars) as the delim. There is a test case in `str_util_test.cc` to show that.

yongtang committed:
Convert SplitUTF8 to return Status

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @drpngx  for the review. The PR has been rebased and updated. Please take a look.

drpngx commented:
Jenkins, test this please.

drpngx committed:
skipEmpty -> skip_empty as per style guide

drpngx committed:
Put error conditions in cold section

drpngx commented:
Sounds fair. Note that since we have it in `str_util`, it's no longer sufficient to check at the single call site, since there might be a different part of the code calling us now.

Jenkins, test this please.

yongtang committed:
Fix `PREDICT_FALSE` -> `TF_PREDICT_FALSE`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @drpngx for the help. I updated the PR with `PREDICT_FALSE` -> `TF_PREDICT_FALSE`.

drpngx commented:
Thanks! Jenkins, test this please.

On Sep 16, 2017 12:29 PM, "Yong Tang" <notifications@github.com> wrote:

> Thanks @drpngx <https://github.com/drpngx> for the help. I updated the PR
> with PREDICT_FALSE -> TF_PREDICT_FALSE.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/12971#issuecomment-329989638>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AT_SbYkV-UiGo-VoZyVbE8EGyp9Kaz4Vks5sjCGNgaJpZM4PTdqy>
> .
>

asimshankar commented on tensorflow/core/ops/string_ops.cc:
```diff
@@ -203,6 +203,7 @@ REGISTER_OP("StringSplit")
     .Output("values: string")
     .Output("shape: int64")
     .Attr("skip_empty: bool = true")
+    .Attr("encoding: string = ''")
```
Please also include a description of what this is in the documentation string below

yongtang replied:
Thanks @asimshankar. The docs has been updated.

asimshankar commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -416,6 +417,86 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
   }
 }
 
+Status SplitUTF8(StringPiece text, const string& delim,
```
Isn't this function invoked only when `delim` is the empty string? 
I realize that you're perhaps adding a more general function, but would it make sense to not have this function in `str_util.h` but instead inline a function to split into UTF8 characters when the delimiter is empty in the anonymous namespace in `kernels/string_split_op.cc`?

yongtang replied:
@asimshankar I updated the PR so that it is possible to have a non empty `delim` when encoding is `utf8`. However, because of the limitation of the current ops definition, only one char (instead of 1-4 chars in utf8 case) is allowed.

asimshankar commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -85,7 +98,9 @@ class StringSplitOp : public OpKernel {
-      std::vector<string> parts = Split(input_vec(i), delimiter, skip_empty_);
+      std::vector<string> parts;
+      OP_REQUIRES_OK(
+          ctx, Split(input_vec(i), delimiter, skip_empty_, encoding_, &parts));
```
If `encoding_ == "utf8"`, should the kernel fail if the delimiter is not valid UTF8?

yongtang replied:
Thanks @asimshankar. The PR has been updated so that delimiter will be checked for properly encoding.

asimshankar reviewed (commented):
From API review: A few comments. The observation that the `encoding` attribute is only relevant when the delimiter is the empty string is a bit surprising.

Not sure: But would it make sense to instead have a separate op that splits a string into UTF8 characters instead of adding to this op?

yongtang committed:
Add additional test case.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@asimshankar Thanks for the review. The PR has been updated. Currently, the implementation of this PR is to:

In case an UTF8 encoding is used, then either an empty string or a one character UTF8 (actually ASCII as there is only one char) is allowed to be specified as a delimiter. Properly encoding is validated now as well.

The reason for only allowing one char in UTF8 case is that, the `StringSplit` OP was defined such that only one char is allowed. If we expand the ops to take more than one chars, I think backward-compatibility may not be maintained.

For that it might make sense to introduce another OP instead of adding additional `encoding` attribute to `StringSplit`. I can update the PR if the other option is preferred. Please let me know.

sb2nov commented:
@asimshankar ping

asimshankar commented:
@yongtang : A separate op sounds preferable as it will be the more complete solution (like wanting to split on a multi-byte character). Let's do that instead.  Thanks!

yongtang commented:
@asimshankar Thanks. I will update the PR.

yongtang committed:
Add `string_utf8_split` op.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks for the review @asimshankar. The PR has been updated with a new OP `string_utf8_split` added, which supports multi-bytes delimiter. Please take a look and let me know if there are any issues.

asimshankar commented on tensorflow/core/lib/strings/str_util.h:
```diff
@@ -131,6 +133,18 @@ std::vector<string> Split(StringPiece text, StringPiece delims);
 
+// Split "text" into into a vector of strings where each string
+// in the array is a string of ASII or unicode chars. The `delim`
+// is a string with 0 or 1 chars.
```
Should be stay "`delim` is a single UTF-8-encoded unicode character or the empty string" (to avoid any confusion, since `delim.size()` can be more than 1)?

yongtang replied:
Done.

asimshankar commented on tensorflow/core/ops/string_ops.cc:
```diff
@@ -247,6 +247,57 @@ shape: a length-2 vector of int64 representing the shape of the sparse
+  shape = [2, 3]
+  values = ['hello', 'world', 'a', 'b', 'c']
+
+input: 1-D. Strings to split.
```
UTF-8 encoded strings to split.

yongtang replied:
Done.

asimshankar commented on tensorflow/core/ops/string_ops.cc:
```diff
@@ -247,6 +247,57 @@ shape: a length-2 vector of int64 representing the shape of the sparse
+  values = ['hello', 'world', 'a', 'b', 'c']
+
+input: 1-D. Strings to split.
+delimiter: 1-D. Delimiter characters (bytes), or empty string.
```
Empty string or a single UTF-8 encoded unicode character

yongtang replied:
Done.

asimshankar commented on tensorflow/core/ops/string_ops.cc:
```diff
@@ -247,6 +247,57 @@ shape: a length-2 vector of int64 representing the shape of the sparse
   of tokens in a single input entry.
 )doc");
 
+REGISTER_OP("StringUTF8Split")
```
Nit: Can we call this `StringSplitUTF8` instead?

yongtang replied:
Done.

asimshankar commented on tensorflow/core/ops/string_ops.cc:
```diff
@@ -247,6 +247,57 @@ shape: a length-2 vector of int64 representing the shape of the sparse
+  values = ['hello', 'world', 'a', 'b', 'c']
+
+input: 1-D. Strings to split.
+delimiter: 1-D. Delimiter characters (bytes), or empty string.
```
I didn't quite follow why `delimiter` is a 1-D tensor. Shouldn't it be a string scalar?

asimshankar commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -122,6 +122,84 @@ class StringSplitOp : public OpKernel {
   bool skip_empty_;
 };
 
+class StringUTF8SplitOp : public OpKernel {
```
IIUC, the implementation of these two kernels is almost identical, save for the different function call to split (and the vall to validating UTF8 of the delimiter).

Could we refactor the kernel implementation so that a bulk of the code is shared? Something along the lines of:

```c++
class StringSplitOpBase : public OpKernel {
    ...
    Status ValidateDelimiter(const string& delimiter) = 0;
    Status Split(const string& input, const string& delimiter, vector<string>* parts) = 0;
}

class StrignSplitOp : public StringSplitOpBase {
  Status ValidateDelimiter(const string& delimiter) { return Status::OK(); }
  Status Split(const string& input, const string& delimiter, vector<string>* parts) {
    return str_util::Split(...);
  }  
};

class StringSplitUTF8Op : public StringSplitOp {
  Status ValidateDelimiter(const string& delimiter) { return str_util::ValidUTF8(delimiter); }
  Status Split(...) { return str_util::SplitUTF8(...); }
};
```

That way it's easy to keep a bulk of the logic in sync?

yongtang replied:
Thanks @asimshankar. The class has been split as suggested.

asimshankar commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -416,6 +417,119 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
+        len = 4;
+      } else {
+        result->clear();
+        return errors::InvalidArgument("Invalid UTF8 encoding at position of ",
```
Replace "position" with "byte" in the error messages to make it clear that the position refers to the byte and not the character?

yongtang replied:
Done.

asimshankar commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -416,6 +417,119 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
+        if (delim == "") {
+          // If delim is "" then always advance
+          result->emplace_back(entry);
+        } else if (delim != entry) {
```
Could this be made a little more compact by separating out the length determination from the emplacement? Something like:

```c++
size_t char_start = 0;
size_t len = 0;
for (size_t i = 0; i < text.size(); ++i) {
  if (char_start == i) {
    // Determine byte length of next character.
    const char c = text[i];
    if (c & 0x80 == 0x00) {
     len = 1;
    } else if (c & 0xE0 == 0xC0) {
      len = 2;
    } else if (c & 0xF0 == 0xE0) {
      len = 3;
    } else if (c & 0XF8 == 0xF0) {
      len = 4;
   } else {
    result->clear();
    return errors::InvalidArgument("Invalid UTF-8 encoding at byte ", i);
   }
   // Validate all intermediate bytes are valid UTF-8
   if (TF_PREDICT_FALSE((len > 1 && (text[i] & 0xC0) != 0x80)) {
      /*  return error */
   }
   // Reached the end of a character
   if (char_start + len == i + 1) {
     const StringPiece entry = text.substr(char_start, len).ToString();
     char_start += len;
     if (delim == "") {
       result->emplace_back(entry);
     } else if (delim == entry) {
        advance = true;
     } else if (advance) {
        advance = false;
        result->emplace_back(entry);
    } else {
        result->back() = result->back() + entry;
   }
}
```

yongtang replied:
Thanks. Done.

asimshankar replied:
Hmm...didn't see any change here. Perhaps a forgotten commit?

asimshankar replied:
Hmm...I didn't see much of a change. By switching things around a bit we won't have to treat the `len == 1` case differently from the others repeat the blocks of code that add to `result`. Something like this:

```c++
Status SplitUTF8(StringPiece text, const string& delim, std::vector<string>* result) {
  result->clear();
  size_t char_start = 0, char_len = 0;
  bool advance = true;
  for (size_t i = 0; i < text.size(); ++i) {
    if (char_start == i) {
      Status s = UTF8CharNumBytes(text, &char_len);
      if (!s.ok()) {
        result->clear();
        return errors::InvalidArgument("Invalid UTF8 encoding at byte ", i, ": ", s.error_message());
      }
    }
    // Validate all intermediate bytes are valid UTF-8
    if (TF_PREDICT_FALSE((len > 1 && (text[i] & 0xC0) != 0x80)) {
       return errors::InvalidArgument("Invalid UTF8 encoding at byte ", i);
    }
    // Reached the end of a character
    if (char_start + char_len == i + 1) {
      const StringPiece entry = text.substr(char_start, char_len).ToString();
      char_start += len;
      if (delim == "") {
        result->emplace_back(entry);
      } else if (delim == entry) {
         advance = true;
      } else if (advance) {
         advance = false;
         result->emplace_back(entry);
      } else {
         result->back() = result->back() + entry;
      }
    }
  }
  if (TF_PREDICT_FALSE(char_start < text.size())) {
    return errors::InvalidArgument("Invalid UTF8 encoding, incomplete last character");
  }
  return Status::OK();
}
```

asimshankar commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -416,6 +417,119 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
+  return Status::OK();
+}
+
+Status ValidUTF8(StringPiece text) {
```
It seems like this function is determining whether `text` is a valid UTF-8 character (as opposed to a valid string). Perhaps call this function so: `ValidUTF8Character` or something?

yongtang replied:
Done.

asimshankar commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -416,6 +417,119 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
+    return Status::OK();
+  }
+  size_t len = 0;
+  if ((text[0] & 0x80) == 0x00) {
```
You could factor this out into a helper function that both this and `Split` use:

```c++
static Status UTF8CharNumBytes(StringPiece text, size_t* num_bytes);
```

yongtang replied:
Thanks. Done.

asimshankar commented on tensorflow/python/kernel_tests/string_split_op_test.py:
```diff
@@ -21,6 +21,7 @@
 import numpy as np
 
 from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import errors
```
Do you need to import `errors` or can you use `errors_impl` like the other tests?

asimshankar reviewed (changes requested):
Thanks for the update @yongtang , some suggestions.

yongtang committed:
Simplication and address review feedback.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @asimshankar for the detailed review. The PR has been updated. Please take a look and let me know if there are any issues.

asimshankar commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -45,19 +45,16 @@ std::vector<string> Split(const string& str, const string& delimiter,
+                       std::vector<string>* parts) = 0;
 
-  void Compute(OpKernelContext* ctx) override {
+  void Compute(OpKernelContext* ctx) {
```
Why was the "override" keyword removed?

yongtang replied:
Added back. Thanks for notice that.

asimshankar commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -118,10 +117,50 @@ class StringSplitOp : public OpKernel {
+  explicit StringSplitUTF8Op(OpKernelConstruction* context)
+      : StringSplitOpBase(context) {
+    // The skip_empty is not optional (unlike StringSplitOp)
+    OP_REQUIRES_OK(context, context->GetAttr("skip_empty", &skip_empty_));
```
I'm probably missing something, but I don't see `skip_empty_` being used by `StringSplitUTF8Op`. How is it having an effect?

yongtang replied:
The PR has been updated with `skip_empty_`.

asimshankar commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -118,10 +117,50 @@ class StringSplitOp : public OpKernel {
   }
 
- private:
+ protected:
```
Let's make `skip_empty_` a private member of the subclasses. Doesn't seem like it us used in the base class and it is only set by the subclass constructors, so it will be cleaner to move it down there?

yongtang replied:
Done.

asimshankar commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -416,6 +417,118 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
+  for (size_t i = 0; i < text.size(); ++i) {
+    if (off == i) {
+      size_t num_bytes = 0;
+      Status s = UTF8CharNumBytes(text.substr(off, 4), &num_bytes);
```
Why `text.substr(off, 4)` here?
That might overflow `text` no? And we could just as well provide `text` (i.e, without `substr`)?

asimshankar commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -416,6 +417,118 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
+  } else if ((text[0] & 0xF8) == 0xF0) {
+    *num_bytes = 4;
+  } else {
+    return errors::InvalidArgument("Invalid UTF8 encoding at position of 0");
```
Hmm...this "position 0" isn't adding much value. Perhaps make this function return a `bool` instead and fill in the status message with offset information at the caller?

yongtang replied:
Done.

asimshankar commented on tensorflow/python/kernel_tests/string_split_op_test.py:
```diff
@@ -145,6 +145,56 @@ def testStringSplitWithNoSkipEmpty(self):
+    strings = [b"\xE6\x82\xA8\xE5\xA5\xBD", b"\xE4\xB8\x96\xE7\x95\x8C"]
+
+    with self.test_session() as sess:
+      tokens = string_ops.string_split_utf8(strings, delimiter="")
```
I don't see any tests for `skip_empty=True`. Also, it appears that `skip_empty` isn't handled. If you'd rather get rid of the `skip_empty` op attribute altogether, that's fine. If not, let's make it work :)

asimshankar replied:
Ping?

yongtang replied:
@asimshankar The tests have been updated with test cases covered when skip_empty=False. Ln 160-208.

asimshankar reviewed (changes requested):
Thanks for the update @yongtang - though now I see that `skip_empty` isn't being handled. And some other suggestions...

yongtang commented:
Thanks @asimshankar for the review. Will updated the PR shortly.

yongtang committed:
Address skip_empty and misc.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @asimshankar for the review. The PR has been updated with review comments addressed. Please take a look.

asimshankar commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -416,6 +417,99 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
+      } else if (delim == entry) {
+        advance = true;
+        if (!skip_empty) {
+          if (TF_PREDICT_FALSE(char_start == 0)) {
```
IIUC, this is handling the case when `skip_empty` is true and the delimiter is the first or last character. But what about cases where there are multiple consecutive delimiters? See comment about testing (and perhaps duplicating/reusing `testStringSplitWithNoSkipEmpty` with this new op)

yongtang replied:
Thanks @asimshankar. My previous interpretation was not right. I have updated the PR so that now it will match python behavior in case `skip_empty=False`, i.e.,
```
>>> "##a##b##c##".split("#")
['', '', 'a', '', 'b', '', 'c', '', '']
>>> 
```

yongtang committed:
Add test cases where multiple consecutive delimieters exist

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@asimshankar Thanks for the review. The PR has been updated. Please take a look.

martinwicke commented:
@asimshankar can you look at the new changes?

asimshankar commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -416,6 +417,103 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
+          // ['', '', 'a', '', 'b', '', 'c', '', '']
+          if (result->size() == 0 || result->back() != "") {
+            result->emplace_back("");
+          } else if (result->size() == 1) {
```
I didn't quite follow the logic here (in particulary while `result->size() == 1` is any more special that `result->size() == 2` etc.). For example, let's say the the call is:

```
SplitUTF8("###a#", '#')
```

then I think this will end up with: `{ "", "", "a", ""}` instead of three empty strings before "a".
If that is correct, probably worth adding the unittest that would have caught the bug here.

I think this can be simplified a bit by keeping a boolean indicating whether the last character was a delimiter and using that instead. Something like:

```c++
} else if (delim == entry) {
  if (last_char_was_a_delimiter && !skip_empty) {
    result->emplace_back("");
  }
  last_char_was_a_delimiter = true;
} else if (last_char_was_a_delimiter) {
  last_char_was_a_delimiter = false;
  result->emplace_back(entry.ToString());
} else {
  result->back() += enty.ToString();
}
```

yongtang replied:
@asimshankar I think some special handling of begin and end may still be needed. For example, in python:
```python
Python 2.7.10 (default, Feb  7 2017, 00:08:15) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> "###a#".split("#")
['', '', '', 'a', '']
>>> ",".split(",")
['', '']
>>> 
```

The first case ` "###a#".split("#")` returns 3 empty chars before `a` and the second case will return 2 chars.

I updated the PR to try to match the above situation. Please take a look and see if it fit.

asimshankar commented on tensorflow/core/lib/strings/str_util_test.cc:
```diff
@@ -417,4 +417,33 @@ TEST(StringReplace, EmptyStringReplaceAll) {
+            str_util::SplitUTF8("\xE2\x28\xA1", "", &result));
+
+  EXPECT_EQ(errors::InvalidArgument("Not enough characters for UTF8 encoding"),
+            str_util::SplitUTF8("\xE6\x82", "", &result));
```
Perhaps also include the simpler string tests, like the ones in `Split.Basic`?

yongtang replied:
Thanks. The tests with `Split.Basic` has been added to test against `SplitUTF8` as well.

asimshankar reviewed (changes requested):
Sorry, somehow missed my radar.

yongtang committed:
Update SplitUTF8 to handle delimiters

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented on tensorflow/core/kernels/string_split_op.cc:
```diff
@@ -117,11 +116,54 @@ class StringSplitOp : public OpKernel {
+    skip_empty_ = true;
+    // By default skip_empty_ is true. We only get the value from attr if it is
+    // available, so that it is backward compatible.
+    context->GetAttr("skip_empty", &skip_empty_);
```
Thanks @asimshankar. Done.

yongtang commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -416,6 +417,103 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
+      }
+    }
+    // Validate all intermediate bytes are valid UTF-8
+    if (TF_PREDICT_FALSE((char_len > 1) && (char_start != i) &&
```
Thanks. Done.

yongtang commented:
Thanks @asimshankar for the review and help. The PR has been updated. Please take a look.

asimshankar commented on tensorflow/core/lib/strings/str_util_test.cc:
```diff
@@ -417,4 +417,77 @@ TEST(StringReplace, EmptyStringReplaceAll) {
+
+  // ASCII should work as well
+  EXPECT_EQ(Status::OK(), str_util::SplitUTF8("", ",", false, &result));
+  EXPECT_EQ(result.size(), 0);
```
Add a test for an empty string split with `skip_empty = true` as well?

yongtang replied:
Thanks. Done.

asimshankar commented on tensorflow/core/ops/string_ops.cc:
```diff
@@ -247,6 +247,57 @@ shape: a length-2 vector of int64 representing the shape of the sparse
+`delimiter` can be empty, or a string of split characters. If `delimiter` is an
+ empty string, each element of `input` is split into individual single-UTF-8
+ character strings, including splitting of UTF-8 multibyte sequences. Otherwise
+ every character of `delimiter` is a potential split point.
```
Sorry, I should have caught this earlier - the documentation here says that `delimiter` can be a sequence of characters, in which case, it splits on any of them. However, the implementation does not seem to do that.

For the original `StringSplit` op:
- [this test](https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/core/lib/strings/str_util_test.cc#L242), and
- [this test](https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/python/kernel_tests/string_split_op_test.py#L110)
exercise the case where there are multiple delimiter characters specified.

So, long story short, unless I'm missing something - for `StringSplitUTF8`, delimiter cannot be a string of split characters, it has to be a single split character.

Is that correct? Instead of adding the complexity of handling multiple split characters for now (unless you're up for it), I'd change this to expect a single split character, and if the feature ever comes up then we'd allow for multiple.

Sound reasonable?

yongtang replied:
Thanks @asimshankar. Given the original scope of #11399 I think that might be the best option for now. We can always add additional processing later if there is a request. The doc string has been updated.

Thanks for pointing this out and really appreciate that.

yongtang committed:
Address review feedback

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -416,6 +417,100 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
+          result->emplace_back("");
+        }
+        last_char_was_a_delimiter = true;
+        if (i + 1 == text.size() && !skip_empty) {
```
Thanks @asimshankar. That works. The PR has been updated.

yongtang commented on tensorflow/core/lib/strings/str_util_test.cc:
```diff
@@ -423,21 +423,65 @@ TEST(SplitUTF8, Basic) {
+
+  // split("###a#", "#") -> { "", "", "", "a", ""}
+  EXPECT_EQ(Status::OK(), str_util::SplitUTF8("###a#", "#", false, &result));
+  EXPECT_EQ(result.size(), 5);
```
Thanks. The PR has been updated.

yongtang commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -416,6 +417,100 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
+      } else if (last_char_was_a_delimiter) {
+        last_char_was_a_delimiter = false;
+        result->emplace_back(entry.ToString());
+      } else if (result->size() == 0) {
```
@asimshankar This condition might actually be needed. The reason is that the following line 
```
      } else {
        result->back() = result->back() + entry.ToString();
      }
```
might have result empty. So incase result is empty then we will initialize instead of append to `result->back()`.

yongtang commented:
Thanks @asimshankar for the help and review. The PR has been updated. Please take a look.

yongtang commented:
There seems to have failed tests. Will update the PR shortly.

tensorflow-jenkins commented:
FAILURE

tensorflow-jenkins commented:
SUCCESS

tensorflow-jenkins commented:
FAILURE

yongtang commented:
The failure is about Windows and the rest of builds seem pass. I don't have a Windows dev machine setup yet, but will try to take a look and see what I could do to fix it.

martinwicke commented:
@asimshankar can you take another look? I triggered the tests.

asimshankar commented:
I was under the impression that @yongtang was digging into the test failures on Windows. Looking at the failures I see:

```
FAIL: testStringSplitWithUtf8AndEmptyDelimiter (__main__.StringSplitOpTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:/tf_jenkins/workspace/tensorflow-pr-win-cmake-py/tensorflow/python/kernel_tests/string_split_op_test.py", line 221, in testStringSplitWithUtf8AndEmptyDelimiter
    b"\xE4\xB8\x96", b"\xE7\x95\x8C"])
  File "C:\Program Files\Anaconda3\lib\site-packages\tensorflow\python\framework\test_util.py", line 1148, in assertAllEqual
    np.testing.assert_array_equal(a, b)
  File "C:\Program Files\Anaconda3\lib\site-packages\numpy\testing\utils.py", line 871, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "C:\Program Files\Anaconda3\lib\site-packages\numpy\testing\utils.py", line 796, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal

(mismatch 100.0%)
 x: array([b'\x00\x82\xa8', b'\x00\xa5\xbd', b'\x00\xb8\x96', b'\x00\x95\x8c'], dtype=object)
 y: array([b'\xe6\x82\xa8', b'\xe5\xa5\xbd', b'\xe4\xb8\x96', b'\xe7\x95\x8c'], 
      dtype='|S3')
```

Suggesting things about the `dtype` of numpy arrays of strings that I am not well versed in.
Will have to look into it. If anyone has ideas, much appreciated. Otherwise will ask around internally as well.

tensorflowbutler commented:
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

drpngx commented:
@yongtang did you investigate this any further?

tensorflowbutler commented:
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

yongtang committed:
Fix build failure.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Address test failure

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
The PR has been rebased to resolve the merge conflict. The Windows issue may still exist.

I finally managed to find a Windows VM image from Microsoft:
https://developer.microsoft.com/en-us/windows/downloads/virtual-machines

I will spend some time to see if I could fix the Windows issue.

tensorflowbutler commented:
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

yongtang commented:
@asimshankar Sorry for lack of updates.

I was able to find a Windows VM with Visual Studio 2017 preinstalled license  from https://developer.microsoft.com/en-us/windows/downloads/virtual-machines (license will expire in 05/13).

However, it appears that TensorFlow currently does not work well with Visual Studio 2017 build.

I have tried multiple times but still could not get a successful Windows build first (either with cmake or with bazel). Therefore unable to move forward to debug the issue of utf8 string with Windows.

GIven the current status I think it might make sense to close this PR or label it as `stalled` for now. I will give it another try and if I am able to move forward again, will open the new PR.

Thanks for the help during the process and sorry for the inconvenience.

drpngx commented:
@guschmue does anything jump at you?

drpngx commented:
Re-running tests to get the logs out.

guschmue commented:
I have not build for some time but can give it a try.

tensorflowbutler commented:
Nagging Assignee @asimshankar: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @asimshankar: It has been 32 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

yongtang committed:
Update API golden

```
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test
           --update_goldens True
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
I feel really ashamed to not be able to fix the Windows test failure after so long. It shouldn't be very hard.

However, I still couldn't even build the tensorflow from source with a Windows 10 VM machine (and Visual Studio 2017), after another attempt. Without that I just could not debug the issue. (Don't have a Windows machine).

I have created an issue #19583 to list the detailed steps for my failure encountered. Would appreciate any help to make tensorflow build on Windows 10 + VS2017.

drpngx commented:
No worries! It's always hard to debug if you don't have access to a system. For the record, the failures are:

```
======================================================================
FAIL: testStringSplitWithUtf8AndEmptyDelimiter (__main__.StringSplitOpTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "T:/src/github/tensorflow/tensorflow/python/kernel_tests/string_split_op_test.py", line 221, in testStringSplitWithUtf8AndEmptyDelimiter
    b"\xE4\xB8\x96", b"\xE7\x95\x8C"])
  File "C:\Python35\lib\site-packages\tensorflow\python\framework\test_util.py", line 1460, in assertAllEqual
    np.testing.assert_array_equal(a, b, err_msg=msg)
  File "C:\Python35\lib\site-packages\numpy\testing\utils.py", line 813, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "C:\Python35\lib\site-packages\numpy\testing\utils.py", line 739, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Arrays are not equal
(mismatch 100.0%)
 x: array([b'\x00\x82\xa8', b'\x00\xa5\xbd', b'\x00\xb8\x96', b'\x00\x95\x8c'], dtype=object)
 y: array([b'\xe6\x82\xa8', b'\xe5\xa5\xbd', b'\xe4\xb8\x96', b'\xe7\x95\x8c'],
      dtype='|S3')
======================================================================
FAIL: testStringSplitWithUtf8AndNonSkipEmpty (__main__.StringSplitOpTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "T:/src/github/tensorflow/tensorflow/python/kernel_tests/string_split_op_test.py", line 170, in testStringSplitWithUtf8AndNonSkipEmpty
    b"", b"\xE7\x95\x8C"])
  File "C:\Python35\lib\site-packages\tensorflow\python\framework\test_util.py", line 1460, in assertAllEqual
    np.testing.assert_array_equal(a, b, err_msg=msg)
  File "C:\Python35\lib\site-packages\numpy\testing\utils.py", line 813, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "C:\Python35\lib\site-packages\numpy\testing\utils.py", line 739, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Arrays are not equal
(mismatch 50.0%)
 x: array([b'\x00\xa5\xbd', b'', b'', b'\x00\x95\x8c'], dtype=object)
 y: array([b'\xe5\xa5\xbd', b'', b'', b'\xe7\x95\x8c'],
      dtype='|S3')
======================================================================
FAIL: testStringSplitWithUtf8AndSkipEmpty (__main__.StringSplitOpTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "T:/src/github/tensorflow/tensorflow/python/kernel_tests/string_split_op_test.py", line 157, in testStringSplitWithUtf8AndSkipEmpty
    self.assertAllEqual(values, [b"\xE5\xA5\xBD", b"\xE7\x95\x8C"])
  File "C:\Python35\lib\site-packages\tensorflow\python\framework\test_util.py", line 1460, in assertAllEqual
    np.testing.assert_array_equal(a, b, err_msg=msg)
  File "C:\Python35\lib\site-packages\numpy\testing\utils.py", line 813, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "C:\Python35\lib\site-packages\numpy\testing\utils.py", line 739, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Arrays are not equal
(mismatch 100.0%)
 x: array([b'\x00\xa5\xbd', b'\x00\x95\x8c'], dtype=object)
 y: array([b'\xe5\xa5\xbd', b'\xe7\x95\x8c'],
      dtype='|S3')
======================================================================
FAIL: testStringSplitWithUtf8AndUtf8MultiBytesDelimiter (__main__.StringSplitOpTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "T:/src/github/tensorflow/tensorflow/python/kernel_tests/string_split_op_test.py", line 235, in testStringSplitWithUtf8AndUtf8MultiBytesDelimiter
    b"\xE4\xB8\x96\xE7\x95\x8C", b"\xE5\xA5\xBD"])
  File "C:\Python35\lib\site-packages\tensorflow\python\framework\test_util.py", line 1460, in assertAllEqual
    np.testing.assert_array_equal(a, b, err_msg=msg)
  File "C:\Python35\lib\site-packages\numpy\testing\utils.py", line 813, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "C:\Python35\lib\site-packages\numpy\testing\utils.py", line 739, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Arrays are not equal
(mismatch 100.0%)
 x: array([b'\x00\xa5\xbd', b'\x00\xb8\x96\xe7\x95\x8c',
       b'\x00\xb8\x96\xe7\x95\x8c', b'\x00\xa5\xbd'], dtype=object)
 y: array([b'\xe5\xa5\xbd', b'\xe4\xb8\x96\xe7\x95\x8c',
       b'\xe4\xb8\x96\xe7\x95\x8c', b'\xe5\xa5\xbd'],
      dtype='|S6')
```
and the MacOS errors are:
```
[ RUN      ] SplitUTF8.Basic
tensorflow/core/lib/strings/str_util_test.cc:470: Failure
      Expected: result[3]
      Which is: "\0"
To be equal to: "a"
tensorflow/core/lib/strings/str_util_test.cc:479: Failure
      Expected: result[2]
      Which is: "\0"
To be equal to: "a"
tensorflow/core/lib/strings/str_util_test.cc:481: Failure
      Expected: result[4]
      Which is: "\0"
To be equal to: "b"
tensorflow/core/lib/strings/str_util_test.cc:483: Failure
      Expected: result[6]
      Which is: "\0"
To be equal to: "c"
tensorflow/core/lib/strings/str_util_test.cc:495: Failure
      Expected: str_util::Join(result, "|")
      Which is: "\0"
To be equal to: "a"
tensorflow/core/lib/strings/str_util_test.cc:501: Failure
      Expected: str_util::Join(result, "|")
      Which is: "\0|\0|\0"
To be equal to: "a|b|c"
tensorflow/core/lib/strings/str_util_test.cc:505: Failure
      Expected: str_util::Join(result, "|")
      Which is: "\0|||\0||\0|"
To be equal to: "a|||b||c|"
tensorflow/core/lib/strings/str_util_test.cc:508: Failure
      Expected: str_util::Join(result, "|")
      Which is: "\0|\0|\0"
```

drpngx commented:
Basically, it looks like there is something wrong with the logic. It seems to return `\x0` instead of the original substrings.

tensorflowbutler commented:
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

tensorflowbutler commented:
It has been 43 days that this pull-request has stalled. Please create a new pull-request with the requested changes.

marhlder commented:
Is there any progress on this or is it just stalled forever?
